src/shared/extension/context/popup.tsx

```diff
@@ -0,0 +1,97 @@
+import {
+  ReactNode,
+  createContext,
+  useCallback,
+  useEffect,
+  useState,
+} from 'react';
+import debounce from 'lodash/debounce';
+import { MemoryRouter, useLocation, useNavigate } from 'react-router';
+
+import { EXTENSION_POPUP_ROUTE } from 'shared/extension';
+import {
+  onWindowMessage,
+  TOGGLE_EXTENSION_POPUP_VISIBILITY,
+} from 'shared/messaging';
+import { createContextHook } from 'shared/ui';
+
+interface Properties {
+  children: ReactNode;
+}
+
+type ExtensionPopupRoute =
+  (typeof EXTENSION_POPUP_ROUTE)[keyof typeof EXTENSION_POPUP_ROUTE];
+
+interface ExtensionPopupContextValues {
+  isVisible: boolean;
+  navigate: (route: ExtensionPopupRoute) => void;
+  currentRoute: ExtensionPopupRoute;
+  hide: () => void;
+  open: () => void;
+}
+
+const ExtensionPopupContext = createContext<
+  ExtensionPopupContextValues | undefined
+>(undefined);
+
+const InnerExtensionPopupProvider = ({ children }: Properties) => {
+  const [isVisible, setIsVisible] = useState(false);
+
+  const navigate = useNavigate();
+  const location = useLocation();
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const toggleVisibility = useCallback(
+    debounce(() => {
+      setIsVisible((previous) => {
+        return !previous;
+      });
+    }, 50),
+    [],
+  );
```

Comment:
I'm not sure if useDebounce from react-use will work here, as it only calls the passed function when the dependency array changes, and does not create a callable debounced function like debounce from lodash. However, I haven't used it much, so I could be wrong.

We can discuss whether we actually need the debounce here. I added it for two reasons:

1. A double-click makes it appear.
2. In React.StrictMode, scripts are injected twice, which prevents the modal from toggling correctly. I believe this could be fixed by adding removeEventListener in the useEffect cleanup